enhancements/microshift/kubernetes-for-device-edge.md

```diff
@@ -0,0 +1,515 @@
+reboots.
+
+Many of the OpenShift APIs are not relevant for edge device use cases.
+We are starting with the absolute minimal set for now because it will
+be easier to add more APIs later than it will be to remove anything.
+It is meant as a pure runtime tool, and will not be used for
+build/dev-cluster use cases, therefore we will not include Builds,
+BuildConfigs, etc.  MicroShift is not an interactive, multi-user
+environment, therefore we do not include some of the authorization
+APIs such as users, projects, etc.  It doesn't manage OS or
+infrastructure, so we do not include MachineConfig and related APIs.
+MicroShift is a monolith, therefore we do not include any cluster
+configuration APIs.
+
+In addition to the standard APIs included in core Kubernetes, the
+OpenShift APIs supported will be:
+
+* route.openshift.io/v1 for Routes
+* security.openshift.io/v1 for SecurityContextConstraints
+* authorization.openshift.io/v1 for RBAC APIs such as Role,
+  RoleBinding, etc.
+
+In addition to those APIs, we will enable two OpenShift controllers
+related to Ingress:
+
+* openshift.io/ingress-ip
+* openshift.io/ingress-to-route
+
+MicroShift is built downstream of OpenShift to ensure that it includes
+the same versions of the embedded components that are present in
+standalone OpenShift, allowing us to maintain binary compatibility.
+The versions of the embedded components are selected by looking at an
+OpenShift release image, finding the git SHAs for the direct
+dependencies, resolving the secondary dependencies, and vendoring the
+results into the MicroShift git repository. That process needs to be
+as automatic as possible, so we want to avoid carrying any
+MicroShift-specific patches that must be applied after a rebase. The
+version of the containerized components are taken from the same
+release image and the image references are built into the MicroShift
+binary.
+
+A great deal of effort has gone into making OpenShift 4 self-managing,
+especially when it comes to multi-node deployment and upgrades. We do
+not want to reinvent that work, nor do we want to recreate the
+OpenShift 3 model using an outside orchestration tool.  A single-node
+control plane becomes a single point of failure when worker nodes are
+attached, which eliminates the high availability benefits of running
+workloads on Kubernetes. Those statements taken together lead us to
+conclude that MicroShift will only support deployments with 1 node
+acting as both control plane and worker.
+
+Because MicroShift does not reuse the system operators from OpenShift,
+it will not be configured through API resources in the way that
+OpenShift is. Instead, we will use a new configuration file for the
+control plane (`/etc/microshift/config.yaml`) with [limited
+configuration
+options](https://github.com/openshift/microshift/blob/main/pkg/config/config.go). The
+node-side configuration will be done via the standard kubelet
+configuration file.
+
```

Comment:
So, microshift is going to depend on two configuration files. 
 - /etc/micorshift/config.yaml  for control plane. 
 - standard configuration for data plane. 

Single binary referring two different config files? Isn't that more prone to errors?

Reply:
I'm ok not documenting the current implementation as things are likely to change. Just wanted to point out that _today_ kubelet config isn't different from other embedded components' config and not something we considered users should change directly. The rationale was to a) minimize exposed config options and protect from users making incompatible changes and b) avoid the need for configuration workflow (MicroShift starts only based on its config.yaml, generates kubelet and other configs (potentially updating them to the new version), then users would need to modify those generated configs).

Some time ago we tossed around the idea of a generic mechanism to pass extra args from MicroShift's config.yaml through to the embedded components (incl. kubelet) in a way we can filter (allow/deny-list) options and decouples from the mechanics of how configs are passed to components. Recently, we also discussed moving closer to OpenShift's config schemas.

Should we keep the mechanism + format open for now?

Reply:
See 3081a2b1